Azure/sonic-buildimage
# Add Host Account Management (HAM) project

**- Dependencies**
It wasn't clear how to put a dependency between two pull requests, hence I added this section.
This pull request depends on [sonic-swss-common #390](https://github.com/Azure/sonic-swss-common/pull/390) being merged first. 

**- Why I did it**
As described in [AAA Improvements](https://github.com/Azure/SONiC/blob/master/doc/aaa/AAA%20Improvements/AAA%20Improvements.md), there is a need for a central location to manage user accounts. The solution is "Host Account Management" or HAM, which acts as the single source of truth for user accounts. Some of the features of HAM include:
  - Automatically assign Linux user credentials (UID, GID) to RADIUS/TACACS+ user accounts.
  - Allow applications running in containers to retrieve Linux user credentials (UID, GID) from the host.
  - Allow applications running in containers (such as a REST server) to create/modify/delete local user accounts on the host. This is only allowed for users with admin privileges.
  - Fixes several problems related to RADIUS/TACACS+ login (for the full list refer to the document cited above). 

**- How I did it**
Not sure what "How I did it" means. I wrote C/C++ code? 
For a description of how to use HAM, refer to the src/ham/README.md. 

**- How to verify it**
This pull request adds a new building block called HAM. In this pull request we do not hook HAM up to any other components in SONiC. In a second phase, we will hook up RADIUS/TACACS+ to use HAM. In a third phase, the sonic-management-framework will also use HAM. Since HAM is not hooked up to anything in this first phase, the only way to verify that things are in place is through the debug utility "**`hamctl`**". This can be invoked at the bash shell.

**- Which release branch to backport (provide reason below if selected)**
No backport needed. This is a new feature.

**- Description for the changelog**
New Host Account Management feature. A detailed description of HAM can be found in: https://github.com/Azure/SONiC/blob/master/doc/aaa/AAA%20Improvements/AAA%20Improvements.md
